### PR TITLE
Use a light wrapper for the image block

### PR DESCRIPTION
**Pull request by youknowriad**
The biggest challenge here is alignments, and the approach I took is to try to match the frontend as closely as possible. In the frontend, for left, right and centered images, there's an extra "div" that is added as a wrapper.

The frontend also relies on `align*` classNames to style the alignments while the backend used to rely on `data-*` attributes. This consolidates on the classNames.

So this PR is most likely to set a precedent to be followed in terms of alignment and something that might extend to all the other blocks.

**Note**

The focus style when the image is floated is kind of broken but it's not specific to this PR.

**Testing**

 - Test the image alignment in different situations

### CODE REVIEW

File: packages/block-library/src/image/edit.js [hunk 1]
```diff
@@ -497,188 +511,196 @@ export class ImageEdit extends Component {
 			</>
 		);
 
+		const AlignmentWrapper = needsAlignmentWrapper ? Block.div : Fragment;
```
Comment (ellatrix) [inline packages/block-library/src/image/edit.js]: Would it make any sense to always have an alignment wrapper?

Comment (youknowriad) [reply packages/block-library/src/image/edit.js]: Even if we do add it always, we still have specific behavior: If the wrapper is "Fragment" it means the figure needs to become Block.figure.

File: packages/block-library/src/image/edit.js [hunk 2]
```diff
@@ -404,10 +405,12 @@ export class ImageEdit extends Component {
 				src={ url }
 			/>
 		);
+		const needsAlignmentWrapper =
+			[ 'center', 'left', 'right' ].indexOf( align ) !== -1;
```
Comment (ZebulanStanphill) [inline packages/block-library/src/image/edit.js]: This could be simplified to:
```suggestion
			[ 'center', 'left', 'right' ].includes( align );
```

**Comment by ellatrix** [comment]
I rebased this PR (force push).
For easier review: https://github.com/WordPress/gutenberg/pull/20576/files?utf8=✓&diff=unified&w=1

**Comment by ellatrix** [comment]
One small thing I notice is that the movers can't be accessed if the image is full width. That's probably because the data- attributes are no longer there and the position of the toolbar won't be adjusted.

File: packages/block-library/src/image/edit.js (hunk 1 above)
Comment (ellatrix) [reply packages/block-library/src/image/edit.js]: Ah, I see it's similar on the front-end. Sounds good then. :)

File: packages/block-library/src/image/edit.js [hunk 3]
```diff
@@ -497,188 +515,196 @@ export class ImageEdit extends Component {
...
+		const BlockContentWrapper = needsAlignmentWrapper
+			? 'figure'
+			: Block.figure;
 		// Disable reason: Each block can be selected by clicking on it
 		/* eslint-disable jsx-a11y/click-events-have-key-events */
 		return (
 			<>
 				{ controls }
-				<figure className={ classes }>
-					<ImageSize src={ url } dirtynessTrigger={ align }>
```
Comment (ellatrix) [inline packages/block-library/src/image/edit.js]: Would be cool if we cool rewrite this as a hook. Something to explore separately.

Comment (youknowriad) [reply packages/block-library/src/image/edit.js]: Definitely, I actually did something like that a long time ago when I was playing with React hooks before their release https://github.com/WordPress/gutenberg/pull/11161

File: packages/block-library/src/image/edit.js [hunk 4]
```diff
@@ -497,188 +515,196 @@ export class ImageEdit extends Component {
...
-								! imageWidthWithinContainer
-							) {
 								return (
 									<>
 										{ getInspectorControls(
 											imageWidth,
 											imageHeight
 										) }
-										<div style={ { width, height } }>
+										<ResizableBox
```
Comment (ellatrix) [inline packages/block-library/src/image/edit.js]: Same here, I have a feeling that this could be rewritten as a hook in the future. :)

File: packages/block-library/src/image/edit.js (hunk 3 above)
Comment (ellatrix) [reply packages/block-library/src/image/edit.js]: Awesome!

File: packages/block-editor/src/components/block-list/style.scss [hunk 5]
```diff
@@ -278,12 +282,15 @@
 
 	// Wide and full-wide.
 	&[data-align="full"],
-	&[data-align="wide"] {
+	&[data-align="wide"],
+	&.alignfull,
+	&.alignwide {
```
Comment (ellatrix) [inline packages/block-editor/src/components/block-list/style.scss]: Hm, ideally this should be provided by the theme, no? I mean, ideally we shouldn't do any special handling for these classes.

File: packages/block-editor/src/components/block-list/style.scss [hunk 6]
```diff
@@ -554,10 +556,6 @@
 			margin-left: - $block-toolbar-height - $border-width;
 		}
 
-		.block-editor-block-contextual-toolbar[data-align="full"],
-		.block-editor-block-list__breadcrumb[data-align="full"] {
-			margin-left: 0;
```
Comment (ellatrix) [inline packages/block-editor/src/components/block-list/style.scss]: Why is this removed? This ensures that the movers are visible for "full" alignment.

File: packages/block-editor/src/components/block-list/style.scss [hunk 7]
```diff
@@ -388,11 +395,6 @@
 		left: $block-padding;
 		right: $block-padding;
 	}
-
-	&[data-align="full"] > .block-editor-block-list__insertion-point {
-		left: 0;
-		right: 0;
```
Comment (ellatrix) [inline packages/block-editor/src/components/block-list/style.scss]: Makes sense, this looks like a dead CSS rule to me.

**Review by ellatrix — APPROVED** [review]
Looks good. I just think this CSS rule should be restored: https://github.com/WordPress/gutenberg/pull/20576/files#r387703250
